airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence2.java

```diff
@@ -0,0 +1,1911 @@
+import io.airbyte.config.StandardSync.Status;
+import io.airbyte.config.StandardSyncOperation;
+import io.airbyte.config.StandardSyncOperation.OperatorType;
+import io.airbyte.config.StandardSyncState;
+import io.airbyte.config.StandardWorkspace;
+import io.airbyte.config.State;
+import io.airbyte.db.Database;
+import io.airbyte.db.ExceptionWrappingDatabase;
+import io.airbyte.db.instance.configs.jooq.enums.ActorType;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConnectorSpecification;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jooq.DSLContext;
+import org.jooq.JSONB;
+import org.jooq.Record;
+import org.jooq.Record1;
+import org.jooq.Result;
+import org.jooq.impl.DSL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DatabaseConfigPersistence2 implements ConfigPersistence {
+
+  private final ExceptionWrappingDatabase database;
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseConfigPersistence2.class);
+  private final DatabaseConfigPersistence oldConfigPersistence;
+  private boolean useNewPersistence;
+
+  public DatabaseConfigPersistence2(final Database database) {
+    this.oldConfigPersistence = new DatabaseConfigPersistence(database);
+    this.database = new ExceptionWrappingDatabase(database);
+    this.useNewPersistence = false;
+  }
+
+  public ValidatingConfigPersistence withValidation() {
+    return new ValidatingConfigPersistence(this);
+  }
+
+  private boolean useNewPersistence() throws IOException {
+    if (useNewPersistence) {
+      LOGGER.info("Using new Persistence");
+      return true;
+    }
+    useNewPersistence = database.query(ctx -> ctx.fetchExists(select()
+        .from("information_schema.tables")
+        .where(DSL.field("table_name").eq(WORKSPACE.getName())
+            .and(DSL.field("table_schema").eq("public")))));
+    if (useNewPersistence) {
+      LOGGER.info("Using new Persistence");
+    } else {
+      LOGGER.info("Using old Persistence");
+    }
+    return useNewPersistence;
+  }
+
+  @Override
+  public <T> T getConfig(AirbyteConfig configType, String configId, Class<T> clazz)
+      throws ConfigNotFoundException, JsonValidationException, IOException {
+    if (!useNewPersistence()) {
+      return oldConfigPersistence.getConfig(configType, configId, clazz);
+    }
+
+    if (configType == ConfigSchema.STANDARD_WORKSPACE) {
+      return (T) getStandardWorkspace(configId);
+    } else if (configType == ConfigSchema.STANDARD_SOURCE_DEFINITION) {
+      return (T) getStandardSourceDefinition(configId);
```

Comment:
i wonder if you can save your self some duplication by piggybacking off of the list queries. e.g. getStandardWorkspaceWithMetadata() optionally took in a config id or you split up `getStandardWorkspaceWithMetadata` to expose the underlying query? there a few ways you might be able to do it.